### PR TITLE
[Kernels] Migrate sampling to WebGPU

### PR DESCRIPTION
**Pull request by akaashrp**
This PR completely replaces CPU sampling with WebGPU sampling. Based on the code in https://github.com/mlc-ai/mlc-llm/blob/main/cpp/serve/sampler/gpu_sampler.cc, the ```logprobs``` flag does not affect the kernels being called. Specifically, branching only occurs based on whether or not we wish to use ```top_p```. In the case where ```top_p``` is not set, we would need to use ```multinomial_sampling_func``` / ```parallel_sampling_from_prob```, which is currently not possible due to the use of ```i8s```. Most models have a default ```top_p``` value, and I set ```top_p = 1.0``` when this is not the case, which allows removal of the else branch.

Performance Comparison (decode tokens/s) with v0.2.79: Compared performance for "canonical" flows averaged across 10 runs with the following configuration:
```
const TEST_PROMPT = "Explain, in at least 100 words, why the phrase 'less errors' is considered non-standard in formal English. Include references to grammatical number agreement, comparative determiners, and prescriptive vs. descriptive norms.";
const reply0 = await engine.chat.completions.create({
    messages: [{ role: "user", content: TEST_PROMPT }],
    n: 1,
    temperature: 0,
    max_tokens: 1024,
    logprobs: false,
});
```
Specifically:
- No logit_bias
- No logitProcessor
- No frequency, presence, and repetition penalties (since the flow in v0.2.79 applies frequency/presence and repetition penalties separately and not together)
- No logprobs (since this is handled differently in v0.2.79—see above)

1. Phi-3.5-mini-instruct-q4f16_1-MLC: 31.765 (v0.2.79) vs 31.754 (this PR)
2. Qwen3-0.6B-q4f16_1-MLC: 39.681 (v0.2.79) vs 40.825 (this PR)
3. Llama-3-8B-Instruct-q4f16_1-MLC: 18.932 (v0.2.79) vs 18.939 (this PR)
4. gemma-2-9b-it-q4f16_1-MLC: 13.642 (v0.2.79) vs 13.863 (this PR)

Notes:
1. The minimal performance improvement is likely due to kernel launch overheads. Specifically, we need to call three kernels to perform sampling (fsoftmaxWithTemperature, fargsortProbs, fSampleWithTopP).
2. Currently, the path corresponding to logprobs takes place on CPU, though a call to the ```sampler_take_probs``` kernel can likely replace this. I will take a deeper look at this in the near future.
3. Added a TODO for exploring using ```multinomial_sampling_func``` for case where ```top_p``` is not explicitly set.
4. Setting ```top_p = 1.0``` by default might run into some correctness issues since masking in relax uses < instead of <= (https://github.com/apache/tvm/blob/main/python/tvm/relax/frontend/nn/op.py#L2795).

### CODE REVIEW

File: src/llm_chat.ts [hunk 1]
```diff
@@ -1284,18 +1317,52 @@ export class LLMChatPipeline {
...
-        this.getTokenLogprob(sampledToken, top_logprobs!),
-      );
+      sampledToken = sampledTokensHost.toArray()[0];
+
+      if (top_logprobs! > 0) {
+        this.tokenLogprobArray.push(
+          this.getTokenLogprob(sampledToken, top_logprobs!),
+        );
+      }
     } else {
```
Comment (CharlieFRuan) [inline src/llm_chat.ts]: quick question: why we cannot use GPU sample kernel when we `logprobs` is False?

Comment (akaashrp) [reply src/llm_chat.ts]: IIRC, the flow for the logprobs False case involves invoking `_attach_multinomial_sampling_func` / `parallel_sampling_from_prob`, which contains i8s that are not supported by WGSL / WebGPU yet. I experimented with enabling some experimental flags at the beginning of the relevant kernels, but I wasn't able to get these to work. One thing I haven't tried yet though is replacing the int8s with some other supported datatype in line 131 here: https://github.com/apache/tvm/blob/26db8bfd7e527198f43f3cc379f404c7513a82ef/python/tvm/relax/backend/gpu_generic/sampling.py#L131C1-L132C1.

Comment (akaashrp) [reply src/llm_chat.ts]: My bad, I had tried that change before. There is an explicit check for the buffer type being boolean / i8, so compilation fails. For completeness, I also tried enabling `chromium_experimental_subgroup_matrix` in WGSL kernels to support i8 on WebGPU, but it is not a valid GPUFeatureName. I also tried enabling `subgroups`, but still ran into the i8 error. Still, what you said makes sense, and I will add a TODO so I can revisit this in more depth at a later stage.

File: src/llm_chat.ts [hunk 2]
```diff
@@ -1271,11 +1302,13 @@ export class LLMChatPipeline {
     // If logprobs, need the actual distribution via softmax, otherwise directly sample from logits
     const sampleBegin = performance.now();
     let sampledToken: number;
-    if (logprobs) {
+    let sampledTokensDevice: tvmjs.Tensor;
+    if (logprobs && _hasValue(top_p)) {
```
[the remark this answers is not in the document]
Comment (akaashrp) [reply src/llm_chat.ts]: Agreed. Based on our discussion offline (summarized here for reference), I've removed the else branch (corresponding to sampling on CPU). 

Based on the code in https://github.com/mlc-ai/mlc-llm/blob/main/cpp/serve/sampler/gpu_sampler.cc, the `logprobs` flag does not affect the kernels being called. Specifically, branching only occurs based on whether or not we wish to use `top_p`. In the case where `top_p` is not set, we would need to use ```multinomial_sampling_func``` / ```parallel_sampling_from_prob```, which is currently not possible as described above. Most models have a default `top_p` value, and I set `top_p = 1.0` when this is not the case, which allows removal of the else branch. 

Note: Currently, the path corresponding to ```logprobs``` takes place on CPU, though this can likely be replaced by a call to the ```sampler_take_probs``` kernel. I will take a deeper look at this in the near future.
